src/safeposix/syscalls/fs_calls.rs

```diff
@@ -26,158 +59,225 @@ impl Cage {
-                            FILEOBJECTTABLE.entry(inodenum)
-                        {
-                            let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
-                            vac.insert(interface::openfile(sysfilename, f.size).unwrap());
-                            // use existing file size
+                            // If the existing file type is a socket, error is thrown as socket type files are not supported by open_syscall
+                            Inode::Socket(_) => {
+                                return syscall_error(Errno::ENXIO, "open", "file is a UNIX domain socket");
+                            }
                         }
 
-                        size = f.size;
-                        f.refcount += 1;
+                        // The file object of size 0, associated with the existing inode number is inserted into the FileDescriptorTable associated with the cage using the guard lock.
+                        let _insertval =
+                            fdoption.insert(File(self._file_initializer(inodenum, flags, size)));
                     }
-                    Inode::Dir(ref mut f) => {
-                        size = f.size;
-                        f.refcount += 1;
-                    }
-                    Inode::CharDev(ref mut f) => {
-                        size = f.size;
-                        f.refcount += 1;
-                    }
-                    Inode::Socket(_) => {
-                        return syscall_error(Errno::ENXIO, "open", "file is a UNIX domain socket");
+
+                    // Case 3: When neither the file directory nor the parent directory exists
+                    (None, None) => {
+                        // O_CREAT flag is used to create a file if it doesn't exist.
+                        // If this flag is not present, then a file can not be created and error is returned.
+                        if 0 == (flags & O_CREAT) {
+                            return syscall_error(
+                                Errno::ENOENT,
+                                "open",
+                                "tried to open a file that did not exist, and O_CREAT was not specified",
+                            );
+                        }
+                        // O_CREAT flag is set but the path doesn't exist, so return an error with a different message string.
+                        return syscall_error(Errno::ENOENT, "open", "a directory component in pathname does not exist or is a dangling symbolic link");
                     }
                 }
 
-                let _insertval =
-                    fdoption.insert(File(self._file_initializer(inodenum, flags, size)));
+                // Once all the updates are done, the file descriptor value is returned
+                fd 
+            },
+            // If the file descriptor is invalid, the return value is always an error with value -23 (ENFILE).
+            -23 => {
```

Comment:
can you use the Errno ENUM here instead of hardcoding it. so that its clear what we are expecting out of the function. and maybe the order of erroring out can be different like this can be checked before the happy case. so we error out quicker.

Reply:
Updated the check with the Enum value and also changed the ordering for it.